File: investigation/probe/mod/src/Job.cs
using System.Collections.Generic;

namespace DGMProbe
{
    // The job the runner writes (runner/job.ts mirrors these shapes; keep the two in step).
    // Game days count as the game does: day 1 starts at 00:00, a new game starts at day 1 + 4/24.

    public class Job
    {
        public int Version;
        public string RunId;
        public string CreatedAt;
        public JobSettings Settings = new JobSettings();
        public List<JobMap> Maps = new List<JobMap>();
    }

    public class JobSettings
    {
        public float Speed = 99f;
        public bool LowGraphics = true;
        public float HeartbeatSeconds = 2f;
        public float MapTimeoutSeconds = 1800f;
        public bool QuitWhenDone = true;
    }

    public class Cycle
    {
        public int TemperateDays;
        public string Hazard = "drought";
        public int HazardDays;
    }

    public class Pose
    {
        public string Id;
        // "look": our 3D view's orbit camera (view world: x east, y up, z = -north), reproduced exactly.
        // "game": the same numbers, but the game's own field of view is kept.
        // "current": the camera exactly as the game has it (the new-game view), no numbers used.
        public string Kind = "game";
        public float[] Target = new float[3];
        public float Yaw;
        public float Pitch;
        public float Distance;
        public float FovY = 40f;
        public int Width = 1280;
        public int Height = 800;
        public string LookCapture;
    }

    public class Moment
    {
        public string Id;
        public double Day;
        public bool Snapshot;
        public List<string> Shots = new List<string>();
    }

    public class ProbeAction
    {
        public double Day;
        public string Kind;
        public string Template;
        public List<int[]> Tiles;
    }

    public class JobMap
    {
        public string Id;
        public string Title;
        public string MapFile;
        public string Faction = "Folktails";
        public string Mode = "Normal";
        public List<Cycle> Cycles = new List<Cycle>();
        public double EndDay;
        // Real seconds this map may take (0: the job's default).
        public float TimeoutSeconds;
        public List<int[]> Tiles = new List<int[]>();
        public double SampleHours = 1;
        public List<Moment> Moments = new List<Moment>();
        public List<ProbeAction> Actions = new List<ProbeAction>();
        public List<Pose> Poses = new List<Pose>();
    }
}

File: investigation/probe/runner/job.ts
// The job file the mod reads (job.json in the folder given with -dgmprobeHome: C:\dgm-probe) and the
// results it writes.
// The C# side mirrors these shapes (mod/src/Job.cs, mod/src/Results.cs); keep the two in step.

/** Game time as the game counts it: day 1 starts at 00:00 of the first day, and a new game starts at
 *  04:00, so the first moment of a game is day 1 + 4/24. Hazards start at 00:00 of a day. */
export type GameDay = number;
export const NEW_GAME_DAY: GameDay = 1 + 4 / 24;

export type Hazard = 'drought' | 'badtide';

/** One weather cycle: temperate days, then the hazard. The mod imposes each cycle as it starts. */
export interface Cycle {
  temperateDays: number;
  hazard: Hazard;
  hazardDays: number;
}

/**
 * A camera pose. `current`: the camera exactly as the game has it (its opening view). `game`: the numbers
 * below with the game's own field of view.
 * `look`: our 3D view's orbit camera, in the view's world (x east, y up, z = −north), vertical field
 * of view `fovY` degrees and an image of `width` × `height`; the mod renders the same frame.
 */
export interface Pose {
  id: string;
  kind: 'look' | 'game' | 'current';
  target: [number, number, number];
  yaw: number;
  pitch: number;
  distance: number;
  fovY: number;
  width: number;
  height: number;
  /** The Map look capture this pose reproduces, relative to docs/map-look/after/. */
  lookCapture?: string;
}

export interface Moment {
  id: string;
  day: GameDay;
  /** A whole-map record: water, soil and plants on every tile. */
  snapshot: boolean;
  shots: string[];
}

export interface Action {
  day: GameDay;
  /** Remove every entity of `template` standing on one of `tiles` (all of them without tiles), as a
   *  finished demolition would. */
  kind: 'deleteEntities';
  template: string;
  tiles?: [number, number][];
}

export interface JobMap {
  id: string;
  title: string;
  /** Absolute path of the .timber the mod starts. */
  mapFile: string;
  faction: 'Folktails' | 'IronTeeth';
  mode: 'Easy' | 'Normal' | 'Hard';
  cycles: Cycle[];
  /** Stop the game at this day. */
  endDay: GameDay;
  /** Real seconds this map may take before the mod gives up on it. */
  timeoutSeconds: number;
  /** Tiles whose water and soil are recorded every `sampleHours`. */
  tiles: [number, number][];
  sampleHours: number;
  moments: Moment[];
  actions: Action[];
  poses: Pose[];
}

export interface JobSettings {
  /** The game speed to hold (the UI's fastest is 3); the mod lowers it if frames get too slow. */
  speed: number;
  lowGraphics: boolean;
  heartbeatSeconds: number;
  /** Real seconds a map may take before the mod gives up on it and moves on. */
  mapTimeoutSeconds: number;
  quitWhenDone: boolean;
}

export interface Job {
  version: 1;
  runId: string;
  createdAt: string;
  settings: JobSettings;
  maps: JobMap[];
}

// ---------------------------------------------------------------------------------------- results

export interface TileSample {
  x: number;
  y: number;
  /** Every water column on the tile, bottom up: floor, depth, contamination. */
  columns: [number, number, number][];
  moisture: number;
  soilContamination: number;
}

export interface SampleRow {
  day: GameDay;
  tick: number;
  weather: string;
  tiles: TileSample[];
}

export interface EntityRecord {
  id: string;
  template: string;
  x: number;
  y: number;
  z: number;
  orientation: string;
  /** Plants: living, dead, dry, flooded, contaminated; growth 0..1. */
  plant?: { dead: boolean; dry: boolean; flooded: boolean; contaminated: boolean; growth: number; diedDay?: GameDay; cause?: string };
  /** Water and badwater sources: specified and current strength, contamination. */
  source?: { specified: number; current: number; contamination: number };
}

export interface MapSnapshot {
  momentId: string;
  day: GameDay;
  tick: number;
  weather: string;
  width: number;
  height: number;
  /** Per tile, the top water column (the one a player sees): depth, contamination, its floor. */
  depth: number[];
  contamination: number[];
  floor: number[];
  moisture: number[];
  soilContamination: number[];
  /** The terrain after the game's terrain physics: the top terrain column's ceiling (first air level)
   *  and the number of terrain columns (1 without caves). */
  terrain: number[];
  terrainColumns: number[];
  /** Tiles with more than one water column (caves, overhangs): x, y and every column. */
  layered: { x: number; y: number; columns: [number, number, number][] }[];
  plants: EntityRecord[];
  sources: EntityRecord[];
}

export interface Shot {
  momentId: string;
  pose: string;
  file: string;
  day: GameDay;
}

export interface LogLine {
  realTime: string;
  type: string;
  message: string;
  stack?: string;
}

export interface MapResult {
  runId: string;
  mapId: string;
  title: string;
  mapFile: string;
  status: 'done' | 'failed' | 'timeout';
  failure?: string;
  gameVersion: string;
  modVersion: string;
  startedAt: string;
  endedAt: string;
  /** Real seconds, and the game speed actually held. */
  realSeconds: number;
  ticks: number;
  meanSpeed: number;
  /** The weather as the game ran it: each cycle's temperate length, hazard and hazard length. */
  weather: { cycle: number; temperateDays: number; hazard: string; hazardDays: number }[];
  weatherEvents: { day: GameDay; event: string }[];
  loadingIssues: string[];
  start: { districtCenter: { x: number; y: number; z: number; orientation: string } | null; adults: number; children: number; bots: number };
  entitiesAtStart: EntityRecord[];
  entitiesAtEnd: EntityRecord[];
  plantDeaths: { id: string; template: string; x: number; y: number; day: GameDay; cause: string }[];
  samples: SampleRow[];
  snapshots: string[];
  shots: Shot[];
  log: LogLine[];
  actions: { day: GameDay; kind: string; template: string; removed: number }[];
  notes: string[];
}
